libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,727 @@
+    if (!_active) {
+        return false;
+    }
+
+    // request thorough cleanup
+    if (_thorough_clean_request_ms == 0) {
+        _thorough_clean_request_ms = AP_HAL::millis();
+        if (clean_type != THOROUGH_CLEAN_DEFAULT) {
+            _thorough_clean_type = clean_type;
+        }
+        return false;
+    }
+
+    // check if background thread has completed request
+    if (_thorough_clean_complete_ms == _thorough_clean_request_ms) {
+        _thorough_clean_request_ms = 0;
+        return true;
+    }
+
+    return false;
+}
+
+// cancel request for thorough cleanup
+void AP_SafeRTL::cancel_request_for_thorough_cleanup()
+{
+    _thorough_clean_request_ms = 0;
+}
+
+//
+// Private methods
+//
+
+// add point to end of path (if necessary), returns true on success
+bool AP_SafeRTL::add_point(const Vector3f& point)
+{
+    // get semaphore
+    if (!_path_sem->take_nonblocking()) {
+        log_action(SRTL_ADD_FAILED_NO_SEMAPHORE, point);
+        return false;
+    }
+
+    // check if we have traveled far enough
+    if (_path_points_count > 0) {
+        const Vector3f& last_pos = _path[_path_points_count-1];
+        if (last_pos.distance_squared(point) < sq(_accuracy.get())) {
+            _path_sem->give();
+            return true;
+        }
+    }
+
+    // check we have space in the path
+    if (_path_points_count >= _path_points_max) {
+        _path_sem->give();
+        log_action(SRTL_ADD_FAILED_PATH_FULL, point);
+        return false;
+    }
+
+    // add point to path
+    _path[_path_points_count++] = point;
+    log_action(SRTL_POINT_ADD, point);
+
+    _path_sem->give();
+    return true;
+}
+
+// run background cleanup - should be run regularly from the IO thread
+void AP_SafeRTL::run_background_cleanup()
+{
+    if (!_active) {
+        return;
+    }
+
+    // get semaphore
+    if (!_path_sem->take_nonblocking()) {
+        return;
+    }
+    // local copy of _path_points_count
+    const uint16_t path_points_count = _path_points_count;
+    _path_sem->give();
+
+    // check if thorough cleanup is required
+    if (_thorough_clean_request_ms > 0) {
+        // check if we have already completed the request
+        if (_thorough_clean_complete_ms != _thorough_clean_request_ms) {
+            if (thorough_cleanup(path_points_count, _thorough_clean_type)) {
+                // record completion
+                _thorough_clean_complete_ms = _thorough_clean_request_ms;
+            }
+        }
+        // we do not perform any further detection or cleanup until the requester acknowledges
+        // they have what they need by setting _thorough_clean_request_ms back to zero
+        return;
+    }
+
+    // ensure clean complete time is zero
+    _thorough_clean_complete_ms = 0;
+
+    // check if path array is nearly full, if yes we should do a routine cleanup (i.e. remove 10 points)
+    bool path_nearly_full = path_points_count >= MAX(_path_points_max - SAFERTL_CLEANUP_START_MARGIN, 0);
```

Comment:
path_points_count would need to be refetched after the thorough cleanup may have been run as it may have changed.